packages/runtime/src/state/react-builder-preview.ts

```diff
@@ -402,15 +404,50 @@ function startPollingElementFromPoint(): ThunkAction<() => void, State, unknown,
   }
 }
 
-export function initialize(): ThunkAction<() => void, State, unknown, Action> {
-  return dispatch => {
+export interface IMessageChannel {
+  postMessage(message: any, transferables?: Transferable[]): void
+  dispatchBuffered(): void
+}
+
+function registerBuilderComponents(state: State, messageChannel: IMessageChannel) {
+  const componentsMeta = getComponentsMeta(state)
+  componentsMeta.forEach((componentMeta, componentType) => {
+    const propControllerDescriptors = getComponentPropControllerDescriptors(state, componentType)
+
+    if (propControllerDescriptors != null) {
+      const [serializedControls, transferables] = serializeControls(propControllerDescriptors)
+
+      messageChannel.postMessage(
+        registerBuilderComponent(componentType, componentMeta, serializedControls),
+        transferables,
+      )
+    }
+  })
+
+  return () => {
+    componentsMeta.forEach((_, componentType) => {
+      messageChannel.postMessage(unregisterBuilderComponent(componentType))
+    })
+  }
+}
+
+export function initialize(
+  channel: IMessageChannel,
+): ThunkAction<() => void, State, unknown, Action> {
+  return (dispatch, getState) => {
+    const state = getState()
     const stopMeasuringElements = dispatch(startMeasuringElements())
     const stopMeasuringDocumentElement = dispatch(startMeasuringDocumentElement())
     const stopHandlingFocusEvent = dispatch(startHandlingFocusEvents())
     const unlockDocumentScroll = dispatch(lockDocumentScroll())
     const stopHandlingPointerMoveEvent = dispatch(startHandlingPointerMoveEvent())
     const stopPollingElementFromPoint = dispatch(startPollingElementFromPoint())
+    const unregisterBuilderComponents = registerBuilderComponents(state, channel)
```

Comment:
Should we make `registerBuilderComponents` a thunk to be consistent so we can do here `dispatch(registerBuilderComponents(channel))` and let it get `state` via `getState` itself?

Reply:
I considered it, but the extra indirection didn't seem to be worth it to me. I can kind of see why you went for decomposing thunks into more thunks here, though, as opposite to just going with a simpler `startMeasuringElements(dispatch)`, will do the same.

Reply:
Realized we actually don't need the channel to register/unregister here, see my other comment.

Reply:
Yeah, I definitely asked myself a lot whether it was useful to make everything a thunk. I opted for it because it seemed like a robust abstraction of "do this thing and I don't care whether it depends on dispatch or state, I just care about the _action_". It's an easy thing to evolve, though. So I don't have a super strong opinion here and am happy to go in whatever direction you prefer.